File: src/python_doctor/checks/resilience.py
from __future__ import annotations

"""Resilience-focused static checks."""

import ast

from .. import project
from ..models import DoctorIssue

def check_bare_except_pass() -> list[DoctorIssue]:
    """except Exception: pass silently swallows errors — at minimum log them."""
    issues: list[DoctorIssue] = []
    for filepath in project.own_python_files():
        try:
            source = filepath.read_text()
            tree = ast.parse(source)
        except Exception:
            continue
        for node in ast.walk(tree):
            if not isinstance(node, ast.ExceptHandler):
                continue
            # Check if the handler body is just `pass` (or `pass` with a comment)
            if len(node.body) == 1 and isinstance(node.body[0], ast.Pass):
                # Check if there's a comment justification on the same line
                line_idx = node.lineno - 1
                lines = source.splitlines()
                has_comment = False
                # Check the except line and the pass line for comments
                for check_line in range(
                    max(0, line_idx), min(len(lines), (node.body[0].end_lineno or node.lineno) + 1)
                ):
                    if "#" in lines[check_line]:
                        has_comment = True
                        break
                if not has_comment:
                    issues.append(
                        DoctorIssue(
                            check="resilience/bare-except-pass",
                            severity="warning",
                            message="except: pass silently swallows errors without logging or comment",
                            path=str(filepath.relative_to(project.REPO_ROOT)),
                            category="Resilience",
                            help="Add logger.debug/warning or a # comment explaining why it's safe to ignore.",
                            line=node.lineno,
                        )
                    )
    return issues

def check_reraise_without_context() -> list[DoctorIssue]:
    """except handlers that re-raise without adding context are noise.

    AST-based detection: finds except handlers whose last statement is a bare
    ``raise`` (or ``raise e`` where ``e`` is the caught exception) AND the handler
    does no useful work (no logging, no cleanup, no variable assignment beyond
    the exception itself). If you catch just to re-raise, remove the try/except.
    If you need cleanup, add context with ``raise NewError(...) from exc``.
    """
    issues: list[DoctorIssue] = []
    for filepath in project.own_python_files():
        try:
            source = filepath.read_text()
            tree = ast.parse(source)
        except Exception:
            continue
        for node in ast.walk(tree):
            if not isinstance(node, ast.ExceptHandler):
                continue
            if not node.body:
                continue
            # Check if last statement is a bare raise
            last_stmt = node.body[-1]
            is_bare_raise = isinstance(last_stmt, ast.Raise) and last_stmt.exc is None
            is_identity_raise = False
            if isinstance(last_stmt, ast.Raise) and last_stmt.exc is not None:
                # `raise e` where e is the caught exception name
                if isinstance(last_stmt.exc, ast.Name) and last_stmt.exc.id == (node.name or ""):
                    is_identity_raise = True
            if not (is_bare_raise or is_identity_raise):
                continue
            # Check if there's any useful work before the raise
            preceding = node.body[:-1]
            has_useful_work = False
            for stmt in preceding:
                # Logging calls, assignments, function calls = useful work
                if isinstance(stmt, (ast.Assign, ast.AugAssign, ast.AnnAssign)):
                    has_useful_work = True
                    break
                if isinstance(stmt, ast.Expr) and isinstance(stmt.value, ast.Call):
                    has_useful_work = True
                    break
                if isinstance(stmt, (ast.If, ast.For, ast.While, ast.With, ast.Try)):
                    has_useful_work = True
                    break
            if has_useful_work:
                continue
            # Check for noqa comment
            lines = source.splitlines()
            if node.lineno <= len(lines) and "# noqa" in lines[node.lineno - 1]:
                continue
            issues.append(
                DoctorIssue(
                    check="resilience/reraise-without-context",
                    severity="warning",
                    message="except handler re-raises without adding context — remove the try/except or add info",
                    path=str(filepath.relative_to(project.REPO_ROOT)),
                    category="Resilience",
                    help="Either remove the try/except (it's noise) or use `raise NewError(...) from exc`.",
                    line=node.lineno,
                )
            )
    return issues


def check_exception_swallowed_silently() -> list[DoctorIssue]:
    """Detect except blocks that swallow exceptions without logging or re-raising.

    Patterns caught:
    - ``except Exception: pass``  (already caught by bare-except-pass for bare except)
    - ``except Exception: return <value>`` with no logging
    - ``except Exception as e:`` with body that never references ``e`` and has no logging

    These silently hide failures, making debugging extremely difficult.
    """
    _LOG_CALLS = {"logger", "logging", "log"}
    issues: list[DoctorIssue] = []
    for filepath in project.own_python_files():
        try:
            source = filepath.read_text()
            tree = ast.parse(source)
        except Exception:
            continue
        lines = source.splitlines()

        for node in ast.walk(tree):
            if not isinstance(node, ast.ExceptHandler):
                continue
            # Only flag 'except Exception' (not bare except, which is caught elsewhere)
            if not (isinstance(node.type, ast.Name) and node.type.id == "Exception"):
                continue
            # Skip if has noqa
            if node.lineno <= len(lines) and "# noqa" in lines[node.lineno - 1]:
                continue

            body = node.body
            if not body:
                continue

            # Check if exception variable is referenced in body
            exc_name = node.name  # e.g. 'e' in 'except Exception as e'
            body_source = ast.dump(ast.Module(body=body, type_ignores=[]))

            has_logging = False
            has_raise = False
            refs_exc = False

            for child in ast.walk(ast.Module(body=body, type_ignores=[])):
                # Check for logging calls
                if isinstance(child, ast.Call) and isinstance(child.func, ast.Attribute):
                    obj = child.func.value
                    if isinstance(obj, ast.Name) and obj.id in _LOG_CALLS:
                        has_logging = True
                    # Also check for logger.xxx(... exc_info=True)
                    if isinstance(obj, ast.Name):
                        for kw in child.keywords:
                            if kw.arg == "exc_info":
                                has_logging = True
                # Check for raise
                if isinstance(child, ast.Raise):
                    has_raise = True
                # Check if exception variable is referenced
                if exc_name and isinstance(child, ast.Name) and child.id == exc_name:
                    refs_exc = True

            # Swallowed = no logging AND no re-raise AND exception variable unused
            if not has_logging and not has_raise:
                # Only flag if: pass, return, or exc variable never used
                is_just_pass = len(body) == 1 and isinstance(body[0], ast.Pass)
                is_just_return = len(body) == 1 and isinstance(body[0], ast.Return)
                exc_unused = exc_name and not refs_exc

                if is_just_pass or is_just_return or exc_unused:
                    issues.append(DoctorIssue(
                        check="resilience/exception-swallowed",
                        severity="warning",
                        message="except Exception block swallows error without logging or re-raising",
                        path=str(filepath.relative_to(project.REPO_ROOT)),
                        category="Resilience",
                        help="Add logger.exception() or logger.warning(..., exc_info=True) to preserve debugging context.",
                        line=node.lineno,
                    ))
    return issues

def check_broad_except_no_context() -> list[DoctorIssue]:
    """Detect 'except Exception' that logs without exc_info context.

    Pattern: ``except Exception: logger.warning("something")`` without
    ``exc_info=True`` — the exception object is lost, making post-mortem
    debugging impossible. The logging call looks like it handles the error
    but actually discards the traceback.
    """
    issues: list[DoctorIssue] = []
    for filepath in project.own_python_files():
        try:
            source = filepath.read_text()
            tree = ast.parse(source)
        except Exception:
            continue
        lines = source.splitlines()

        for node in ast.walk(tree):
            if not isinstance(node, ast.ExceptHandler):
                continue
            if not (isinstance(node.type, ast.Name) and node.type.id == "Exception"):
                continue
            if node.lineno <= len(lines) and "# noqa" in lines[node.lineno - 1]:
                continue

            body = node.body
            has_raise = any(isinstance(c, ast.Raise) for c in ast.walk(ast.Module(body=body, type_ignores=[])))
            if has_raise:
                continue

            # Find logging calls without exc_info
            for child in ast.walk(ast.Module(body=body, type_ignores=[])):
                if not isinstance(child, ast.Call):
                    continue
                func = child.func
                if not isinstance(func, ast.Attribute):
                    continue
                obj = func.value
                if not (isinstance(obj, ast.Name) and obj.id in ("logger", "logging", "log")):
                    continue
                if func.attr not in ("warning", "warn", "info", "debug"):
                    continue
                # Check if exc_info=True is passed
                has_exc_info = any(
                    kw.arg == "exc_info" and isinstance(kw.value, ast.Constant) and kw.value.value is True
                    for kw in child.keywords
                )
                if not has_exc_info:
                    # Check if the exception variable is used as an argument
                    exc_name = node.name
                    refs_exc = False
                    if exc_name:
                        for arg in child.args:
                            if isinstance(arg, ast.Name) and arg.id == exc_name:
                                refs_exc = True
                            elif isinstance(arg, ast.JoinedStr):
                                for val in arg.values:
                                    if isinstance(val, ast.FormattedValue) and isinstance(val.value, ast.Name) and val.value.id == exc_name:
                                        refs_exc = True
                    if not refs_exc:
                        issues.append(DoctorIssue(
                            check="resilience/broad-except-no-context",
                            severity="warning",
                            message=f"except Exception logs via logger.{func.attr}() but discards traceback",
                            path=str(filepath.relative_to(project.REPO_ROOT)),
                            category="Resilience",
                            help="Add exc_info=True to the logging call or include the exception variable in the message.",
                            line=child.lineno,
                        ))
                    break  # Only flag once per except block
    return issues


__all__ = [
    "check_bare_except_pass",
    "check_broad_except_no_context",
    "check_exception_swallowed_silently",
    "check_reraise_without_context",
]

File: rust/doctor_core/src/rules/resilience.rs
use rustpython_parser::ast::{self, Expr, Ranged, Stmt};

use crate::{Issue, ModuleIndex, RuleSelection};
use crate::ast_helpers::*;

pub(crate) fn collect_resilience_issues(
    module: &ModuleIndex,
    suite: &ast::Suite,
    rules: &RuleSelection,
) -> Vec<Issue> {
    if !module.source.contains("except") {
        return Vec::new();
    }
    let mut issues = Vec::new();
    walk_suite_stmts(suite, &mut |stmt| {
        let handlers = match stmt {
            Stmt::Try(node) => Some(&node.handlers),
            Stmt::TryStar(node) => Some(&node.handlers),
            _ => None,
        };
        let Some(handlers) = handlers else { return };
        for handler in handlers {
            let ast::ExceptHandler::ExceptHandler(handler) = handler;
            let handler_line = module.line_for_offset(handler.range.start().to_usize());

            // bare-except-pass: except handler body is just `pass` with no comment
            if rules.bare_except_pass
                && handler.body.len() == 1
                && matches!(handler.body[0], Stmt::Pass(_))
            {
                let pass_line = module.line_for_offset(handler.body[0].range().start().to_usize());
                let mut has_comment = false;
                for check_line in handler_line..=pass_line {
                    if check_line > 0
                        && check_line <= module.lines.len()
                        && module.lines[check_line - 1].raw.contains('#')
                    {
                        has_comment = true;
                        break;
                    }
                }
                if !has_comment {
                    issues.push(Issue {
                        check: "resilience/bare-except-pass",
                        severity: "warning",
                        category: "Resilience",
                        line: handler_line,
                        path: module.rel_path.clone(),
                        message: "except: pass silently swallows errors without logging or comment",
                        help: "Add logger.debug/warning or a # comment explaining why it's safe to ignore.",
                    });
                }
            }

            // reraise-without-context
            if rules.reraise_without_context && !handler.body.is_empty() {
                let last_stmt = &handler.body[handler.body.len() - 1];
                let is_bare_raise = matches!(last_stmt, Stmt::Raise(r) if r.exc.is_none());
                let is_identity_raise = if let Stmt::Raise(r) = last_stmt {
                    if let Some(exc) = &r.exc {
                        if let Expr::Name(name) = &**exc {
                            handler.name.as_deref() == Some(name.id.as_str())
                        } else {
                            false
                        }
                    } else {
                        false
                    }
                } else {
                    false
                };
                if is_bare_raise || is_identity_raise {
                    let preceding = &handler.body[..handler.body.len() - 1];
                    let has_useful_work = preceding.iter().any(|s| {
                        matches!(
                            s,
                            Stmt::Assign(_)
                                | Stmt::AugAssign(_)
                                | Stmt::AnnAssign(_)
                                | Stmt::If(_)
                                | Stmt::For(_)
                                | Stmt::While(_)
                                | Stmt::With(_)
                                | Stmt::Try(_)
                        ) || matches!(s, Stmt::Expr(e) if matches!(&*e.value, Expr::Call(_)))
                    });
                    if !has_useful_work
                        && !module.is_rule_suppressed(handler_line, "resilience/reraise-without-context")
                    {
                        issues.push(Issue {
                            check: "resilience/reraise-without-context",
                            severity: "warning",
                            category: "Resilience",
                            line: handler_line,
                            path: module.rel_path.clone(),
                            message: "except handler re-raises without adding context — remove the try/except or add info",
                            help: "Either remove the try/except (it's noise) or use `raise NewError(...) from exc`.",
                        });
                    }
                }
            }

            // exception-swallowed and broad-except-no-context: only for `except Exception`
            let is_except_exception = handler.type_.as_ref().is_some_and(|t| {
                matches!(&**t, Expr::Name(n) if n.id.as_str() == "Exception")
            });
            if !is_except_exception {
                continue;
            }
            if module.is_rule_suppressed(handler_line, "resilience/exception-swallowed")
                && module.is_rule_suppressed(handler_line, "resilience/broad-except-no-context")
            {
                continue;
            }

            let body = &handler.body;
            let exc_name = handler.name.as_deref();
            let mut has_logging = false;
            let mut has_raise = false;
            let mut refs_exc = false;
            let mut log_call_without_context: Option<(usize, &str)> = None;

            walk_suite_exprs(body, &mut |expr| {
                if let Expr::Call(call) = expr {
                    if let Expr::Attribute(func) = &*call.func {
                        if let Expr::Name(obj) = &*func.value {
                            let is_log = matches!(obj.id.as_str(), "logger" | "logging" | "log");
                            if is_log {
                                has_logging = true;
                                let has_exc_info = call.keywords.iter().any(|kw| {
                                    kw.arg.as_deref() == Some("exc_info")
                                        && matches!(&kw.value, Expr::Constant(c) if matches!(c.value, ast::Constant::Bool(true)))
                                });
                                if !has_exc_info
                                    && matches!(func.attr.as_str(), "warning" | "warn" | "info" | "debug")
                                    && log_call_without_context.is_none()
                                {
                                    let mut call_refs_exc = false;
                                    if let Some(exc_n) = exc_name {
                                        for arg in &call.args {
                                            if matches!(arg, Expr::Name(n) if n.id.as_str() == exc_n) {
                                                call_refs_exc = true;
                                            }
                                            if let Expr::JoinedStr(fstr) = arg {
                                                for val in &fstr.values {
                                                    if let Expr::FormattedValue(fv) = val {
                                                        if matches!(&*fv.value, Expr::Name(n) if n.id.as_str() == exc_n)
                                                        {
                                                            call_refs_exc = true;
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                    if !call_refs_exc {
                                        let line =
                                            module.line_for_offset(call.range.start().to_usize());
                                        log_call_without_context =
                                            Some((line, func.attr.as_str()));
                                    }
                                }
                            }
                        }
                    }
                }
            });
            walk_suite_stmts(body, &mut |s| {
                if matches!(s, Stmt::Raise(_)) {
                    has_raise = true;
                }
            });
            if let Some(exc_n) = exc_name {
                walk_suite_exprs(body, &mut |expr| {
                    if matches!(expr, Expr::Name(n) if n.id.as_str() == exc_n) {
                        refs_exc = true;
                    }
                });
            }

            // exception-swallowed: no logging, no raise, and exc unused or just pass/return
            if rules.exception_swallowed
                && !has_logging
                && !has_raise
                && !module.is_rule_suppressed(handler_line, "resilience/exception-swallowed")
            {
                let is_just_pass = body.len() == 1 && matches!(body[0], Stmt::Pass(_));
                let is_just_return = body.len() == 1 && matches!(body[0], Stmt::Return(_));
                let exc_unused = exc_name.is_some() && !refs_exc;
                if is_just_pass || is_just_return || exc_unused {
                    issues.push(Issue {
                        check: "resilience/exception-swallowed",
                        severity: "warning",
                        category: "Resilience",
                        line: handler_line,
                        path: module.rel_path.clone(),
                        message: "except Exception block swallows error without logging or re-raising",
                        help: "Add logger.exception() or logger.warning(..., exc_info=True) to preserve debugging context.",
                    });
                }
            }

            // broad-except-no-context: logging without exc_info
            if rules.broad_except_no_context
                && !has_raise
                && !module.is_rule_suppressed(handler_line, "resilience/broad-except-no-context")
            {
                if let Some((log_line, log_attr)) = log_call_without_context {
                    issues.push(Issue {
                        check: "resilience/broad-except-no-context",
                        severity: "warning",
                        category: "Resilience",
                        line: log_line,
                        path: module.rel_path.clone(),
                        message: Box::leak(
                            format!(
                                "except Exception logs via logger.{}() but discards traceback",
                                log_attr
                            )
                            .into_boxed_str(),
                        ),
                        help: "Add exc_info=True to the logging call or include the exception variable in the message.",
                    });
                }
            }
        }
    });
    issues
}

// ── Security: SQL f-string interpolation ────────────────────────────────

